go-controller/pkg/clustermanager/status_manager/status_manager.go

```diff
@@ -0,0 +1,122 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	ktypes "k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+// fieldManagerName should be different from any zone name
+const fieldManagerName = "cluster-manager"
+
+type typedStatusManager[T Object] struct {
+	informer cache.SharedIndexInformer
+	// cluster-scoped resources should ignore namespace
+	patch func(ctx context.Context, namespace, name string, pt ktypes.PatchType, data []byte, opts metav1.PatchOptions,
+		subresources ...string) (T, error)
+	// updateStatusFunc should handle nil in case of pointer type, and return (updatedObject, needsUpdate).
+	// When possible, return a new object with just the right status set, this makes encoding faster.
+	updateStatusFunc func(obj T) ([]byte, bool)
+}
+
+type Object interface {
+	runtime.Object
+	metav1.Object
+}
+
+func (m *typedStatusManager[T]) updateStatus(oldObjInterface, newObjInterface interface{}) {
+	oldObj, ok := oldObjInterface.(T)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expecting %T but received %T", *new(T), oldObj))
+		return
+	}
+	newObj, ok := newObjInterface.(T)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expecting %T but received %T", *new(T), newObj))
+		return
+	}
+
+	patch, needsUpdate := m.updateStatusFunc(newObj)
+	if !needsUpdate {
+		return
+	}
+
+	patchOptions := metav1.PatchOptions{
+		Force:        utilpointer.Bool(true),
+		FieldManager: fieldManagerName,
+	}
+	_, err := m.patch(context.TODO(), newObj.GetNamespace(), newObj.GetName(), ktypes.ApplyPatchType, patch, patchOptions, "status")
+
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("could not patch: %w", err))
+	}
+}
+
+func (m *typedStatusManager[T]) Start() error {
+	_, err := m.informer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: m.updateStatus,
+		})
+	return err
+}
+
+type StatusManager struct {
+	abpRouteManager *typedStatusManager[*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute]
+}
+
+func NewStatusManager(wf *factory.WatchFactory, ovnClient *util.OVNClusterManagerClientset) *StatusManager {
+	sm := &StatusManager{}
+	if config.OVNKubernetesFeature.EnableMultiExternalGateway {
+		sm.abpRouteManager = &typedStatusManager[*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute]{
+			informer: wf.APBRouteInformer().Informer(),
+			patch: func(ctx context.Context, namespace, name string, pt ktypes.PatchType, data []byte, opts metav1.PatchOptions,
+				subresources ...string) (*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute, error) {
+				return ovnClient.AdminPolicyRouteClient.K8sV1().AdminPolicyBasedExternalRoutes().Patch(ctx, name, pt, data, opts, subresources...)
+			},
+			updateStatusFunc: updateStatusAPBRoute,
+		}
+	}
+	return sm
+}
+
+func (sm *StatusManager) Start() error {
+	klog.Infof("Starting StatusManager with params: abpRouteManager=%v",
```

Comment:
maybe we could make this function more generic by using reflect to get the fields of the struct and iterate through each one?

Reply:
yeah that's easy

Reply:
I don't think it helped enough :/

This code is bound to grow with more statuses being managed and we will keep adding and in-lining more code.

If you want to stick to generics, we could have an interface that could be implemented on a separate file for each type
```
type statusHandler[T any] interface {
    get ...
    patch ...
    ...
}
```

And then another type with the common logic
```
type statusController[T statusHandler[T]] struct {
    statusHandler statusHandler[T]
    informer ...
    ...
}
```

Just an idea

Reply:
I tried something like that, but now I killed lint, because it can't find interface usages. But the code should work

Reply:
I think the lint failure is this https://github.com/dominikh/go-tools/issues/1440
I may need to switch back to inlining instead of using an interface, until the bug is fixed, @jcaamano wdyt?